storage: TestStoreRangeSystemSplits failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/c53084cbd4b2be90789652ffb74b3d0c02930f7c

Parameters:
```
TAGS=
GOFLAGS=-parallel=4
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestStoreRangeSystemSplits PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1239993&tab=buildLog

```
but found:
[/Meta2/System/NodeLiveness /Meta2/System/NodeLivenessMax /Meta2/System/tsd /Meta2/System/"tse" /Meta2/Table/SystemConfigSpan/Start /Meta2/Table/11 /Meta2/Table/12 /Meta2/Table/13 /Meta2/Table/14 /Meta2/Table/15 /Meta2/Table/16 /Meta2/Table/17 /Meta2/Table/18 /Meta2/Table/19 /Meta2/Table/20 /Meta2/Table/21 /Meta2/Table/22 /Meta2/Table/23 /Meta2/Table/24 /Meta2/Table/50 /Meta2/Table/51 /Meta2/Table/52 /Meta2/Table/53 /Meta2/Table/54 /Meta2/Max]
I190413 07:21:11.458383 568978 storage/replica_command.go:252  [split,s1,r25/1:/{Table/54-Max}] initiating a split of this range at key /Table/57 [r26] (zone config)
I190413 07:21:11.459806 568428 storage/queue.go:435  [split,s1,r25/1:/{Table/54-Max},txn=628aa133] rate limited in maybeadd-merge: throttled on async limiting semaphore
E190413 07:21:11.460401 569219 storage/consistency_queue.go:137  [consistencyChecker,s1,r25/1:/{Table/54-Max}] computing own checksum: could not dial node ID 1: no node dialer configured
E190413 07:21:11.460434 569219 storage/queue.go:866  [consistencyChecker,s1,r25/1:/{Table/54-Max}] computing own checksum: could not dial node ID 1: no node dialer configured
I190413 07:21:11.462768 568413 storage/queue.go:435  [split,s1,r25/1:/Table/5{4-7},txn=628aa133] rate limited in maybeadd-replicate: throttled on async limiting semaphore
I190413 07:21:11.648281 567901 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] storage.IntentResolver: cleanup txn records
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190413 07:21:11.648324 567901 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] storage.IntentResolver: cleanup txn records
W190413 07:21:11.648348 569240 storage/intentresolver/intent_resolver.go:820  failed to gc transaction record: could not GC completed transaction anchored at /Local/Range/Table/54/RangeDescriptor: node unavailable; try another peer
--- FAIL: TestStoreRangeSystemSplits (45.22s)
    client_split_test.go:1367: condition failed to evaluate within 45s: expected split keys:
        [/Meta2/System/NodeLiveness /Meta2/System/NodeLivenessMax /Meta2/System/tsd /Meta2/System/"tse" /Meta2/Table/SystemConfigSpan/Start /Meta2/Table/11 /Meta2/Table/12 /Meta2/Table/13 /Meta2/Table/14 /Meta2/Table/15 /Meta2/Table/16 /Meta2/Table/17 /Meta2/Table/18 /Meta2/Table/19 /Meta2/Table/20 /Meta2/Table/21 /Meta2/Table/22 /Meta2/Table/23 /Meta2/Table/24 /Meta2/Table/50 /Meta2/Table/51 /Meta2/Table/52 /Meta2/Table/53 /Meta2/Table/54 /Meta2/Table/57 /Meta2/Max]
        but found:
        [/Meta2/System/NodeLiveness /Meta2/System/NodeLivenessMax /Meta2/System/tsd /Meta2/System/"tse" /Meta2/Table/SystemConfigSpan/Start /Meta2/Table/11 /Meta2/Table/12 /Meta2/Table/13 /Meta2/Table/14 /Meta2/Table/15 /Meta2/Table/16 /Meta2/Table/17 /Meta2/Table/18 /Meta2/Table/19 /Meta2/Table/20 /Meta2/Table/21 /Meta2/Table/22 /Meta2/Table/23 /Meta2/Table/24 /Meta2/Table/50 /Meta2/Table/51 /Meta2/Table/52 /Meta2/Table/53 /Meta2/Table/54 /Meta2/Max]
        goroutine 567901 [running]:
        runtime/debug.Stack(0xa7a358200, 0xc0016a4ff0, 0x34d8a00)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0xa7
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x353ffa0, 0xc001330000, 0xc0016a4fc0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:49 +0x103
        github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeSystemSplits.func2()
        	/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_split_test.go:1334 +0xb12
        github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeSystemSplits(0xc001330000)
        	/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_split_test.go:1367 +0x55d
        testing.tRunner(0xc001330000, 0x2ee4cd8)
        	/usr/local/go/src/testing/testing.go:827 +0xbf
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:878 +0x35c

```

SHA: https://github.com/cockroachdb/cockroach/commits/ee451b4f273e93018969bc0edd34e7105625adaf

Parameters:
```
TAGS=
GOFLAGS=-parallel=4
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestStoreRangeSystemSplits PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1241085&tab=buildLog

```
but found:
[/Meta2/System/NodeLiveness /Meta2/System/NodeLivenessMax /Meta2/System/tsd /Meta2/System/"tse" /Meta2/Table/SystemConfigSpan/Start /Meta2/Table/11 /Meta2/Table/12 /Meta2/Table/13 /Meta2/Table/14 /Meta2/Table/15 /Meta2/Table/16 /Meta2/Table/17 /Meta2/Table/18 /Meta2/Table/19 /Meta2/Table/20 /Meta2/Table/21 /Meta2/Table/22 /Meta2/Table/23 /Meta2/Table/24 /Meta2/Table/50 /Meta2/Table/51 /Meta2/Table/52 /Meta2/Table/53 /Meta2/Table/54 /Meta2/Max]
I190414 05:45:52.080330 516157 storage/replica_command.go:252  [split,s1,r25/1:/{Table/54-Max}] initiating a split of this range at key /Table/57 [r26] (zone config)
I190414 05:45:52.082707 515462 storage/queue.go:435  [consistencyChecker,s1,r25/1:/{Table/54-Max}] rate limited in maybeadd-merge: throttled on async limiting semaphore
E190414 05:45:52.083007 516156 storage/consistency_queue.go:137  [consistencyChecker,s1,r25/1:/{Table/54-Max}] computing own checksum: could not dial node ID 1: no node dialer configured
E190414 05:45:52.083046 516156 storage/queue.go:866  [consistencyChecker,s1,r25/1:/{Table/54-Max}] computing own checksum: could not dial node ID 1: no node dialer configured
I190414 05:45:52.086092 515469 storage/queue.go:435  [split,s1,r25/1:/Table/5{4-7},txn=a2026e9b] rate limited in maybeadd-replicate: throttled on async limiting semaphore
I190414 05:45:52.278945 515336 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] storage.IntentResolver: cleanup txn records
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190414 05:45:52.279051 515336 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] storage.IntentResolver: cleanup txn records
1      [async] intent_resolver_gc_batcher
I190414 05:45:52.279099 515336 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] storage.IntentResolver: cleanup txn records
W190414 05:45:52.279111 516231 storage/intentresolver/intent_resolver.go:820  failed to gc transaction record: could not GC completed transaction anchored at /Local/Range/Table/54/RangeDescriptor: node unavailable; try another peer
--- FAIL: TestStoreRangeSystemSplits (45.23s)
    client_split_test.go:1367: condition failed to evaluate within 45s: expected split keys:
        [/Meta2/System/NodeLiveness /Meta2/System/NodeLivenessMax /Meta2/System/tsd /Meta2/System/"tse" /Meta2/Table/SystemConfigSpan/Start /Meta2/Table/11 /Meta2/Table/12 /Meta2/Table/13 /Meta2/Table/14 /Meta2/Table/15 /Meta2/Table/16 /Meta2/Table/17 /Meta2/Table/18 /Meta2/Table/19 /Meta2/Table/20 /Meta2/Table/21 /Meta2/Table/22 /Meta2/Table/23 /Meta2/Table/24 /Meta2/Table/50 /Meta2/Table/51 /Meta2/Table/52 /Meta2/Table/53 /Meta2/Table/54 /Meta2/Table/57 /Meta2/Max]
        but found:
        [/Meta2/System/NodeLiveness /Meta2/System/NodeLivenessMax /Meta2/System/tsd /Meta2/System/"tse" /Meta2/Table/SystemConfigSpan/Start /Meta2/Table/11 /Meta2/Table/12 /Meta2/Table/13 /Meta2/Table/14 /Meta2/Table/15 /Meta2/Table/16 /Meta2/Table/17 /Meta2/Table/18 /Meta2/Table/19 /Meta2/Table/20 /Meta2/Table/21 /Meta2/Table/22 /Meta2/Table/23 /Meta2/Table/24 /Meta2/Table/50 /Meta2/Table/51 /Meta2/Table/52 /Meta2/Table/53 /Meta2/Table/54 /Meta2/Max]
        goroutine 515336 [running]:
        runtime/debug.Stack(0xa7a358200, 0xc001d01470, 0x34d9620)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0xa7
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x3540bc0, 0xc000788000, 0xc001d01440)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:49 +0x103
        github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeSystemSplits.func2()
        	/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_split_test.go:1334 +0xb12
        github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeSystemSplits(0xc000788000)
        	/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_split_test.go:1367 +0x55d
        testing.tRunner(0xc000788000, 0x2ee5840)
        	/usr/local/go/src/testing/testing.go:827 +0xbf
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:878 +0x35c

```

SHA: https://github.com/cockroachdb/cockroach/commits/d3f704f839ccaef7f10c3af48c78a26d390ae1dc

Parameters:
```
TAGS=
GOFLAGS=-parallel=4
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestStoreRangeSystemSplits PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1241797&tab=buildLog

```
but found:
[/Meta2/System/NodeLiveness /Meta2/System/NodeLivenessMax /Meta2/System/tsd /Meta2/System/"tse" /Meta2/Table/SystemConfigSpan/Start /Meta2/Table/11 /Meta2/Table/12 /Meta2/Table/13 /Meta2/Table/14 /Meta2/Table/15 /Meta2/Table/16 /Meta2/Table/17 /Meta2/Table/18 /Meta2/Table/19 /Meta2/Table/20 /Meta2/Table/21 /Meta2/Table/22 /Meta2/Table/23 /Meta2/Table/24 /Meta2/Table/50 /Meta2/Table/51 /Meta2/Table/52 /Meta2/Table/53 /Meta2/Table/54 /Meta2/Max]
I190415 06:12:17.335996 537101 storage/replica_command.go:252  [split,s1,r25/1:/{Table/54-Max}] initiating a split of this range at key /Table/57 [r26] (zone config)
E190415 06:12:17.337635 537055 storage/consistency_queue.go:137  [consistencyChecker,s1,r25/1:/{Table/54-Max}] computing own checksum: could not dial node ID 1: no node dialer configured
E190415 06:12:17.337670 537055 storage/queue.go:866  [consistencyChecker,s1,r25/1:/{Table/54-Max}] computing own checksum: could not dial node ID 1: no node dialer configured
I190415 06:12:17.339848 536406 storage/queue.go:435  [split,s1,r25/1:/Table/5{4-7},txn=8f18d6a3] rate limited in maybeadd-replicate: throttled on async limiting semaphore
I190415 06:12:17.340197 536394 storage/queue.go:435  [s1,r26/1:/{Table/57-Max}] rate limited in maybeadd-merge: throttled on async limiting semaphore
I190415 06:12:17.528789 536222 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] storage.IntentResolver: cleanup txn records
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190415 06:12:17.528849 536222 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] storage.IntentResolver: cleanup txn records
W190415 06:12:17.528917 536865 storage/intentresolver/intent_resolver.go:820  failed to gc transaction record: could not GC completed transaction anchored at /Local/Range/Table/54/RangeDescriptor: node unavailable; try another peer
--- FAIL: TestStoreRangeSystemSplits (45.22s)
    client_split_test.go:1367: condition failed to evaluate within 45s: expected split keys:
        [/Meta2/System/NodeLiveness /Meta2/System/NodeLivenessMax /Meta2/System/tsd /Meta2/System/"tse" /Meta2/Table/SystemConfigSpan/Start /Meta2/Table/11 /Meta2/Table/12 /Meta2/Table/13 /Meta2/Table/14 /Meta2/Table/15 /Meta2/Table/16 /Meta2/Table/17 /Meta2/Table/18 /Meta2/Table/19 /Meta2/Table/20 /Meta2/Table/21 /Meta2/Table/22 /Meta2/Table/23 /Meta2/Table/24 /Meta2/Table/50 /Meta2/Table/51 /Meta2/Table/52 /Meta2/Table/53 /Meta2/Table/54 /Meta2/Table/57 /Meta2/Max]
        but found:
        [/Meta2/System/NodeLiveness /Meta2/System/NodeLivenessMax /Meta2/System/tsd /Meta2/System/"tse" /Meta2/Table/SystemConfigSpan/Start /Meta2/Table/11 /Meta2/Table/12 /Meta2/Table/13 /Meta2/Table/14 /Meta2/Table/15 /Meta2/Table/16 /Meta2/Table/17 /Meta2/Table/18 /Meta2/Table/19 /Meta2/Table/20 /Meta2/Table/21 /Meta2/Table/22 /Meta2/Table/23 /Meta2/Table/24 /Meta2/Table/50 /Meta2/Table/51 /Meta2/Table/52 /Meta2/Table/53 /Meta2/Table/54 /Meta2/Max]
        goroutine 536222 [running]:
        runtime/debug.Stack(0xa7a358200, 0xc002150090, 0x34d9620)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0xa7
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x3540bc0, 0xc0019baf00, 0xc002150060)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:49 +0x103
        github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeSystemSplits.func2()
        	/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_split_test.go:1334 +0xb12
        github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeSystemSplits(0xc0019baf00)
        	/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_split_test.go:1367 +0x55d
        testing.tRunner(0xc0019baf00, 0x2ee5840)
        	/usr/local/go/src/testing/testing.go:827 +0xbf
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:878 +0x35c

```

SHA: https://github.com/cockroachdb/cockroach/commits/162528d0429389364f07221e7b818925243d7519

Parameters:
```
TAGS=
GOFLAGS=-parallel=4
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestStoreRangeSystemSplits PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1243927&tab=buildLog

```
but found:
[/Meta2/System/NodeLiveness /Meta2/System/NodeLivenessMax /Meta2/System/tsd /Meta2/System/"tse" /Meta2/Table/SystemConfigSpan/Start /Meta2/Table/11 /Meta2/Table/12 /Meta2/Table/13 /Meta2/Table/14 /Meta2/Table/15 /Meta2/Table/16 /Meta2/Table/17 /Meta2/Table/18 /Meta2/Table/19 /Meta2/Table/20 /Meta2/Table/21 /Meta2/Table/22 /Meta2/Table/23 /Meta2/Table/24 /Meta2/Table/50 /Meta2/Table/51 /Meta2/Table/52 /Meta2/Table/53 /Meta2/Table/54 /Meta2/Max]
I190416 06:24:39.170480 575736 storage/replica_command.go:252  [split,s1,r25/1:/{Table/54-Max}] initiating a split of this range at key /Table/57 [r26] (zone config)
E190416 06:24:39.174103 575749 storage/consistency_queue.go:137  [consistencyChecker,s1,r25/1:/{Table/54-Max}] computing own checksum: could not dial node ID 1: no node dialer configured
E190416 06:24:39.174140 575749 storage/queue.go:955  [consistencyChecker,s1,r25/1:/{Table/54-Max}] computing own checksum: could not dial node ID 1: no node dialer configured
I190416 06:24:39.175221 574803 storage/queue.go:470  [split,s1,r25/1:/Table/5{4-7},txn=358c81b3] rate limited in maybeadd-replicate: throttled on async limiting semaphore
I190416 06:24:39.175484 574803 storage/queue.go:470  [s1,r26/1:/{Table/57-Max}] rate limited in maybeadd-merge: throttled on async limiting semaphore
I190416 06:24:39.385106 574745 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] storage.IntentResolver: cleanup txn records
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190416 06:24:39.385140 574745 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] storage.IntentResolver: cleanup txn records
W190416 06:24:39.385185 575758 storage/intentresolver/intent_resolver.go:827  failed to gc transaction record: could not GC completed transaction anchored at /Local/Range/Table/54/RangeDescriptor: node unavailable; try another peer
--- FAIL: TestStoreRangeSystemSplits (65.25s)
    client_split_test.go:1367: condition failed to evaluate within 45s: expected split keys:
        [/Meta2/System/NodeLiveness /Meta2/System/NodeLivenessMax /Meta2/System/tsd /Meta2/System/"tse" /Meta2/Table/SystemConfigSpan/Start /Meta2/Table/11 /Meta2/Table/12 /Meta2/Table/13 /Meta2/Table/14 /Meta2/Table/15 /Meta2/Table/16 /Meta2/Table/17 /Meta2/Table/18 /Meta2/Table/19 /Meta2/Table/20 /Meta2/Table/21 /Meta2/Table/22 /Meta2/Table/23 /Meta2/Table/24 /Meta2/Table/50 /Meta2/Table/51 /Meta2/Table/52 /Meta2/Table/53 /Meta2/Table/54 /Meta2/Table/57 /Meta2/Max]
        but found:
        [/Meta2/System/NodeLiveness /Meta2/System/NodeLivenessMax /Meta2/System/tsd /Meta2/System/"tse" /Meta2/Table/SystemConfigSpan/Start /Meta2/Table/11 /Meta2/Table/12 /Meta2/Table/13 /Meta2/Table/14 /Meta2/Table/15 /Meta2/Table/16 /Meta2/Table/17 /Meta2/Table/18 /Meta2/Table/19 /Meta2/Table/20 /Meta2/Table/21 /Meta2/Table/22 /Meta2/Table/23 /Meta2/Table/24 /Meta2/Table/50 /Meta2/Table/51 /Meta2/Table/52 /Meta2/Table/53 /Meta2/Table/54 /Meta2/Max]
        goroutine 574745 [running]:
        runtime/debug.Stack(0xa7a358200, 0xc001e4fef0, 0x34de560)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0xa7
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x3545fc0, 0xc000f2e100, 0xc001e4fec0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:49 +0x103
        github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeSystemSplits.func2()
        	/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_split_test.go:1334 +0xb12
        github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeSystemSplits(0xc000f2e100)
        	/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_split_test.go:1367 +0x55d
        testing.tRunner(0xc000f2e100, 0x2eea1a8)
        	/usr/local/go/src/testing/testing.go:827 +0xbf
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:878 +0x35c

```

I suspect #36755 fixes this too.

`master`:

```
--- FAIL: TestStoreRangeSystemSplits (45.20s)
    client_split_test.go:1367: condition failed to evaluate within 45s: expected split keys:
        [/Meta2/System/NodeLiveness /Meta2/System/NodeLivenessMax /Meta2/System/tsd /Meta2/System/"tse" /Meta2/Table/SystemConfigSpan/Start /Meta2/Table/11 /Meta2/Table/12 /Meta2/Table/13 /Meta2/Table/14 /Meta2/Table/15 /Meta2/Table/16 /Meta2/Table/17 /Meta2/Table/18 /Meta2/Table/19 /Meta2/Table/20 /Meta2/Table/21 /Meta2/Table/22 /Meta2/Table/23 /Meta2/Table/24 /Meta2/Table/50 /Meta2/Table/51 /Meta2/Table/52 /Meta2/Table/53 /Meta2/Table/54 /Meta2/Table/57 /Meta2/Max]
        but found:
        [/Meta2/System/NodeLiveness /Meta2/System/NodeLivenessMax /Meta2/System/tsd /Meta2/System/"tse" /Meta2/Table/SystemConfigSpan/Start /Meta2/Table/11 /Meta2/Table/12 /Meta2/Table/13 /Meta2/Table/14 /Meta2/Table/15 /Meta2/Table/16 /Meta2/Table/17 /Meta2/Table/18 /Meta2/Table/19 /Meta2/Table/20 /Meta2/Table/21 /Meta2/Table/22 /Meta2/Table/23 /Meta2/Table/24 /Meta2/Table/50 /Meta2/Table/51 /Meta2/Table/52 /Meta2/Table/53 /Meta2/Table/54 /Meta2/Max]
        goroutine 54 [running]:
        runtime/debug.Stack(0xa7a358200, 0xc000d08ed0, 0x33d8ee0)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x9d
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x3498100, 0xc0001d6b00, 0xc000d08ea0)
        	/home/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:49 +0xfa
        github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeSystemSplits.func2()
        	/home/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_split_test.go:1334 +0xc2b
        github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeSystemSplits(0xc0001d6b00)
        	/home/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_split_test.go:1367 +0x5d9
        testing.tRunner(0xc0001d6b00, 0x2e0a598)
        	/usr/local/go/src/testing/testing.go:865 +0xc0
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:916 +0x357
FAIL


ERROR: exit status 1

177 runs completed, 1 failures, over 2m55s
```

that PR (note also how much faster it runs each iteration)

```
51748 runs so far, 0 failures, over 9m35s
```

--- a/pkg/storage/queue.go
+++ b/pkg/storage/queue.go
@@ -226,9 +226,8 @@ type queueConfig struct {
 	maxSize int
 	// maxConcurrency is the maximum number of replicas that can be processed
 	// concurrently. If not set, defaults to 1.
-	maxConcurrency  int
-	addSemSize      int
-	maybeAddSemSize int
+	maxConcurrency       int
+	addOrMaybeAddSemSize int
 	// needsLease controls whether this queue requires the range lease to
 	// operate on a replica. If so, one will be acquired if necessary.
 	needsLease bool
@@ -293,13 +292,12 @@ type baseQueue struct {
 	store  *Store
 	gossip *gossip.Gossip
 	queueConfig
-	incoming    chan struct{} // Channel signaled when a new replica is added to the queue.
-	processSem  chan struct{}
-	addSem      chan struct{} // for .AddAsync
-	maybeAddSem chan struct{} // for .MaybeAddAsync
-	addLogN     log.EveryN    // avoid log spam when addSem, maybeAddSem are maxed out
-	processDur  int64         // accessed atomically
-	mu          struct {
+	incoming         chan struct{} // Channel signaled when a new replica is added to the queue.
+	processSem       chan struct{}
+	addOrMaybeAddSem chan struct{} // for {Maybe,}AddAsync
+	addLogN          log.EveryN    // avoid log spam when addSem, addOrMaybeAddSemSize are maxed out
+	processDur       int64         // accessed atomically
+	mu               struct {
 		syncutil.Mutex                                    // Protects all variables in the mu struct
 		replicas       map[roachpb.RangeID]*replicaItem   // Map from RangeID to replicaItem
 		priorityQ      priorityQueue                      // The priority queue
@@ -326,11 +324,10 @@ func newBaseQueue(
 	if cfg.maxConcurrency == 0 {
 		cfg.maxConcurrency = 1
 	}
-	if cfg.addSemSize == 0 {
-		cfg.addSemSize = 100
-	}
-	if cfg.maybeAddSemSize == 0 {
-		cfg.maybeAddSemSize = 500
+	// NB: addOrmaybeAddSemSize coupled with tight scanner intervals in tests
+	// unfortunately bog down the race build if they are increased too much.
+	if cfg.addOrMaybeAddSemSize == 0 {
+		cfg.addOrMaybeAddSemSize = 20
 	}
 
 	ambient := store.cfg.AmbientCtx
@@ -342,17 +339,16 @@ func newBaseQueue(
 	}
 
 	bq := baseQueue{
-		AmbientContext: ambient,
-		name:           name,
-		impl:           impl,
-		store:          store,
-		gossip:         gossip,
-		queueConfig:    cfg,
-		incoming:       make(chan struct{}, 1),
-		processSem:     make(chan struct{}, cfg.maxConcurrency),
-		addSem:         make(chan struct{}, cfg.maxConcurrency),
-		maybeAddSem:    make(chan struct{}, cfg.maxConcurrency),
-		addLogN:        log.Every(5 * time.Second),
+		AmbientContext:   ambient,
+		name:             name,
+		impl:             impl,
+		store:            store,
+		gossip:           gossip,
+		queueConfig:      cfg,
+		incoming:         make(chan struct{}, 1),
+		processSem:       make(chan struct{}, cfg.maxConcurrency),
+		addOrMaybeAddSem: make(chan struct{}, cfg.addOrMaybeAddSemSize),
+		addLogN:          log.Every(5 * time.Second),
 		getReplica: func(id roachpb.RangeID) (replicaInQueue, error) {
 			repl, err := store.GetReplica(id)
 			if repl == nil || err != nil {
@@ -435,40 +431,63 @@ func (bq *baseQueue) Start(stopper *stop.Stopper) {
 	bq.processLoop(stopper)
 }
 
-// AddAsync asynchronously adds the specified replica to the queue, regardless of the
-// return value of bq.shouldQueue. The replica is added with specified
-// priority. If the queue is too full, the replica may not be added,
-// as the replica with the lowest priority will be dropped. Returns
-// (true, nil) if the replica was added, (false, nil) if the replica
-// was already present, and (false, err) if the replica could not be
-// added for any other reason.
-func (bq *baseQueue) AddAsync(ctx context.Context, repl *Replica, priority float64) {
-	opName := "add-" + bq.name
-	if err := bq.store.stopper.RunLimitedAsyncTask(ctx, opName,
-		bq.addSem, false, /* wait */
-		func(ctx context.Context) {
-			_, _ = bq.addInternal(ctx, repl.Desc(), true, priority)
-		}); err != nil && bq.addLogN.ShouldLog() {
+type baseQueueHelper struct {
+	bq *baseQueue
+}
 
+func (h baseQueueHelper) MaybeAdd(ctx context.Context, repl replicaInQueue, now hlc.Timestamp) {
+	h.bq.maybeAdd(ctx, repl, now)
+}
+
+func (h baseQueueHelper) Add(ctx context.Context, repl replicaInQueue, prio float64) {
+	_, err := h.bq.addInternal(ctx, repl.Desc(), true /* should */, prio)
+	if err != nil && log.V(1) {
+		log.Infof(ctx, "during Add: %s", err)
+	}
+}
+
+type queueHelper interface {
+	MaybeAdd(ctx context.Context, repl replicaInQueue, now hlc.Timestamp)
+	Add(ctx context.Context, repl replicaInQueue, prio float64)
+}
+
+// Async is a more performant substitute for calling AddAsync or MaybeAddAsync
+// when many operations are going to be carried out. It invokes the given helper
+// function in a goroutine if semaphore capacity is available. If the semaphore
+// is not available, the 'wait' parameter decides whether to wait or to return
+// as a noop. Note that if the system is quiescing, fn may never be called in-
+// dependent of the value of 'wait'.
+func (bq *baseQueue) Async(
+	ctx context.Context, opName string, wait bool, fn func(ctx context.Context, h queueHelper),
+) {
+	if log.V(3) {
+		log.InfofDepth(ctx, 2, opName)
+	}
+	opName += " (" + bq.name + ")"
+	if err := bq.store.stopper.RunLimitedAsyncTask(ctx, opName, bq.addOrMaybeAddSem, wait,
+		func(ctx context.Context) {
+			fn(ctx, baseQueueHelper{bq})
+		}); err != nil && bq.addLogN.ShouldLog() {
 		log.Infof(ctx, "rate limited in %s: %s", opName, err)
 	}
 }
 
-// MaybeAddAsync asynchronously adds the specified replica if bq.shouldQueue
-// specifies it should be queued. Replicas are added to the queue using the
-// priority returned by bq.shouldQueue. If the queue is too full, the replica
-// may not be added, as the replica with the lowest priority will be dropped.
+// MaybeAddAsync offers the replica to the queue. The queue will only process a
+// certain number of these operations concurrently, and will drop (i.e. treat as
+// a noop) any additional calls.
 func (bq *baseQueue) MaybeAddAsync(ctx context.Context, repl replicaInQueue, now hlc.Timestamp) {
-	opName := "maybeadd-" + bq.name
-	if err := bq.store.stopper.RunLimitedAsyncTask(
-		ctx, opName,
-		bq.maybeAddSem, false, /* wait */
-		func(ctx context.Context) {
-			bq.maybeAdd(ctx, repl, now)
-		}); err != nil && bq.addLogN.ShouldLog() {
+	bq.Async(ctx, "MaybeAdd", false /* wait */, func(ctx context.Context, h queueHelper) {
+		h.MaybeAdd(ctx, repl, now)
+	})
+}
 
-		log.Infof(ctx, "rate limited in %s: %s", opName, err)
-	}
+// AddAsync adds the replica to the queue. Unlike MaybeAddAsync, it will wait
+// for other operations to finish instead of turning into a noop (because
+// unlikely MaybeAdd, Add is not subject to being called opportunistically).
+func (bq *baseQueue) AddAsync(ctx context.Context, repl replicaInQueue, prio float64) {
+	bq.Async(ctx, "Add", false /* wait */, func(ctx context.Context, h queueHelper) {
+		h.Add(ctx, repl, prio)
+	})
 }
 
 func (bq *baseQueue) maybeAdd(ctx context.Context, repl replicaInQueue, now hlc.Timestamp) {

--- a/pkg/storage/replica.go
+++ b/pkg/storage/replica.go
@@ -43,6 +43,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/storage/storagebase"
 	"github.com/cockroachdb/cockroach/pkg/storage/storagepb"
 	"github.com/cockroachdb/cockroach/pkg/storage/txnwait"
+	"github.com/cockroachdb/cockroach/pkg/util"
 	"github.com/cockroachdb/cockroach/pkg/util/envutil"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/humanizeutil"
@@ -424,6 +425,13 @@ type Replica struct {
 		// not have all the log entries.
 		draining bool
 	}
+
+	// Throttle how often we offer this Replica to the split and merge queues.
+	// We have triggers downstream of Raft that do so based on limited
+	// information and without explicit throttling some replicas will offer once
+	// per applied Raft command, which is silly and also clogs up the queues'
+	// semaphores.
+	splitQueueThrottle, mergeQueueThrottle util.EveryN
 
 	// loadBasedSplitter keeps information about load-based splitting.
 	loadBasedSplitter split.Decider

--- a/pkg/storage/replica_init.go
+++ b/pkg/storage/replica_init.go
@@ -27,12 +27,18 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/storage/stateloader"
 	"github.com/cockroachdb/cockroach/pkg/storage/storagebase"
 	"github.com/cockroachdb/cockroach/pkg/storage/txnwait"
+	"github.com/cockroachdb/cockroach/pkg/util"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
 	"github.com/pkg/errors"
 	"go.etcd.io/etcd/raft"
+)
+
+const (
+	splitQueueThrottleDuration = 5 * time.Second
+	mergeQueueThrottleDuration = 5 * time.Second
 )
 
 func newReplica(rangeID roachpb.RangeID, store *Store) *Replica {
@@ -69,6 +75,9 @@ func newReplica(rangeID roachpb.RangeID, store *Store) *Replica {
 	// replica GC issues, but is a distraction at the moment.
 	// r.AmbientContext.AddLogTagStr("@", fmt.Sprintf("%x", unsafe.Pointer(r)))
 	r.raftMu.stateLoader = stateloader.Make(rangeID)
+
+	r.splitQueueThrottle = util.Every(splitQueueThrottleDuration)
+	r.mergeQueueThrottle = util.Every(mergeQueueThrottleDuration)
 	return r
 }
 

--- a/pkg/storage/replica_proposal.go
+++ b/pkg/storage/replica_proposal.go
@@ -535,6 +535,11 @@ func (r *Replica) handleReplicatedEvalResult(
 		r.store.splitQueue.MaybeAddAsync(ctx, r, r.store.Clock().Now())
 	}
 	if r.store.mergeQueue != nil && needsMergeBySize { // the bootstrap store has a nil merge queue
+		// TODO(tbg): for ranges which are small but protected from merges by
+		// other means (zone configs etc), this is called on every command, and
+		// fires off a goroutine each time. Make this trigger (and potentially
+		// the split one above, though it hasn't been observed to be as
+		// bothersome) less aggressive.
 		r.store.mergeQueue.MaybeAddAsync(ctx, r, r.store.Clock().Now())
 	}
 

--- a/pkg/storage/replica_proposal.go
+++ b/pkg/storage/replica_proposal.go
@@ -531,10 +531,15 @@ func (r *Replica) handleReplicatedEvalResult(
 	r.store.metrics.addMVCCStats(deltaStats)
 	rResult.Delta = enginepb.MVCCStatsDelta{}
 
-	if r.store.splitQueue != nil && needsSplitBySize { // the bootstrap store has a nil split queue
+	// NB: the bootstrap store has a nil split queue.
+	// TODO(tbg): the above is probably a lie now.
+	if r.store.splitQueue != nil && needsSplitBySize && r.splitQueueThrottle.ShouldProcess(timeutil.Now()) {
 		r.store.splitQueue.MaybeAddAsync(ctx, r, r.store.Clock().Now())
 	}
-	if r.store.mergeQueue != nil && needsMergeBySize { // the bootstrap store has a nil merge queue
+
+	// The bootstrap store has a nil merge queue.
+	// TODO(tbg): the above is probably a lie now.
+	if r.store.mergeQueue != nil && needsMergeBySize && r.mergeQueueThrottle.ShouldProcess(timeutil.Now()) {
 		// TODO(tbg): for ranges which are small but protected from merges by
 		// other means (zone configs etc), this is called on every command, and
 		// fires off a goroutine each time. Make this trigger (and potentially

--- a/pkg/storage/replicate_test.go
+++ b/pkg/storage/replicate_test.go
@@ -30,13 +30,12 @@ import (
 func TestEagerReplication(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
-	t.Skip("https://github.com/cockroachdb/cockroach/issues/36663")
-
 	ctx := context.Background()
 	storeCfg := storage.TestStoreConfig(nil /* clock */)
 	// Disable the replica scanner so that we rely on the eager replication code
 	// path that occurs after splits.
 	storeCfg.TestingKnobs.DisableScanner = true
+
 	stopper := stop.NewStopper()
 	defer stopper.Stop(ctx)
 	store := createTestStoreWithConfig(t, stopper, storeCfg)
@@ -46,6 +45,7 @@ func TestEagerReplication(t *testing.T) {
 	// replication cannot succeed).
 	purgatoryStartCount := store.ReplicateQueuePurgatoryLength()
 
+	t.Logf("purgatory start count is %d", purgatoryStartCount)
 	// Perform a split and check that there's one more range in the purgatory.
 
 	key := roachpb.Key("a")

--- a/pkg/storage/store.go
+++ b/pkg/storage/store.go
@@ -1682,8 +1682,12 @@ func (s *Store) systemGossipUpdate(sysCfg *config.SystemConfig) {
 		log.Event(ctx, "computed initial metrics")
 	})
 
+	// We'll want to offer all replicas to the split and merge queues. Be a little
+	// careful about not spawning too many individual goroutines.
+
 	// For every range, update its zone config and check if it needs to
 	// be split or merged.
+	now := s.cfg.Clock.Now()
 	newStoreReplicaVisitor(s).Visit(func(repl *Replica) bool {
 		key := repl.Desc().StartKey
 		zone, err := sysCfg.GetZoneConfigForKey(key)
@@ -1694,8 +1698,12 @@ func (s *Store) systemGossipUpdate(sysCfg *config.SystemConfig) {
 			zone = config.DefaultZoneConfigRef()
 		}
 		repl.SetZoneConfig(zone)
-		s.mergeQueue.MaybeAddAsync(ctx, repl, s.cfg.Clock.Now())
-		s.splitQueue.MaybeAddAsync(ctx, repl, s.cfg.Clock.Now())
+		s.splitQueue.Async(ctx, "gossip update", true /* wait */, func(ctx context.Context, h queueHelper) {
+			h.MaybeAdd(ctx, repl, now)
+		})
+		s.mergeQueue.Async(ctx, "gossip update", true /* wait */, func(ctx context.Context, h queueHelper) {
+			h.MaybeAdd(ctx, repl, now)
+		})
 		return true // more
 	})
 }

